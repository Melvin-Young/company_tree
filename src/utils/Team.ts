export interface IStaff {
  id: number,
  name: string,
  team: string
}

export interface ITeam {
  name: string,
  getParent: () => ITeam,
  setParent: (parent: ITeam) => void,
  getChildren: () => Map<string, ITeam>,
  getChild: (child: string) => ITeam,
  addChild: (child: ITeam) => Map<string, ITeam>,
  hasChild: (child: string) => boolean,
  getStaff: () => Map<number, IStaff>,
  addStaff: (staff: IStaff) => Map<number, IStaff>,
  getNestedStaff: () => Map<number, IStaff>,
  addNestedStaff: (staff: IStaff) => Map<number, IStaff>
}

export default class Team implements ITeam {
  public name: string;
  private parent: ITeam;
  private children: Map<string, ITeam>;
  private staff: Map<number, IStaff>;
  private nestedStaff: Map<number, IStaff>;

  constructor(name: string) {
    if (!name) {
      throw new Error('Name is required');
    }
    this.name = name;
    this.parent = {} as ITeam;
    this.children = new Map();
    this.staff = new Map();
    this.nestedStaff = new Map();
  }

  public getParent(): ITeam {
    return this.parent;
  }

  public setParent(parent: ITeam): void {
    this.parent = parent;
  }

  public getChildren(): Map<string, ITeam> {
    return this.children;
  }

  public getChild(child: string): ITeam {
    return this.children.get(child) as ITeam;
  }

  public addChild(child: ITeam): Map<string, ITeam> {
    return this.children.set(child.name, child);
  }

  public hasChild(child: string): boolean {
    return this.children.has(child);
  }

  public getStaff(): Map<number, IStaff> {
    return this.staff;
  }

  public addStaff(staff: IStaff): Map<number, IStaff> {
    return this.staff.set(staff.id, staff);
  }

  public getNestedStaff(): Map<number, IStaff> {
    return this.nestedStaff;
  }

  public addNestedStaff(staff: IStaff): Map<number, IStaff> {
    return this.nestedStaff.set(staff.id, staff);
  }
}