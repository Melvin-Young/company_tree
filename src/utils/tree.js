export default class Tree {
  constructor(rootNode) {
    this.root = rootNode;
  }

  addTeam(newNode, parentNodes) {
    if(!newNode || Object.getOwnPropertyNames(newNode).length === 0) {
      throw new Error('Team node is invalid')
    }

    let parent = this.search(parentNodes);
    if (parent) {
      newNode.parent = parent;
      parent.children.set(newNode.name, newNode);
    } else {
      return {
        message: 'The parent team specified does not exist',
        node: newNode
      }
    }
  }

  search(parentNodes = [], currentNode = this.root) {
    let rootName = parentNodes.shift();
    if (!parentNodes.length && rootName === this.root.name) {
      return this.root;
    } 

    let currentSearch = parentNodes.shift();
   
    while (currentSearch && currentNode.children.has(currentSearch)) {
      currentNode = currentNode.children.get(currentSearch);
      if (!parentNodes.length) {
        return currentNode;
      }
      currentSearch = parentNodes.shift();
    }

    return false;
  }

  bfsSearch(name, currentNode = this.root) {
    //Function takes ambiguous search string and returns all results it can find
    let result = [];
    let searchQueue = [];
  
    searchQueue.push(currentNode);
  
    while (searchQueue && searchQueue.length) {
      let nodeToCheck = searchQueue.shift();
      nodeToCheck.name === name && result.push(nodeToCheck);
      nodeToCheck.children.forEach(child => searchQueue.push(child));
    }  
    return result;
  }
}
