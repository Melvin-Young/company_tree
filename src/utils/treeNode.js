 export default class {
  constructor(name) {
    if (!name || typeof name !== 'string') {
      throw new Error('Name is required and must be a string');
    }
    this.name = name;
    this.parent = null;
    this.children = new Map();
    this.isMarked = false;
  }
}