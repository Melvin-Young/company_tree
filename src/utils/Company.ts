import Team, {IStaff, ITeam} from './Team';

export interface ICompany {
  companyName: string,
  getRoot: () => ITeam,
  addTeam: (team: ITeam, ancestors: string[]) => Map<string, ITeam>,
  search: (ancestors: string[]) => ITeam,
  addMemberToTeam: (pathToTeam: string[], staff: IStaff) => Map<number, IStaff>,
  getStaff: (pathToTeam: string[]) => Map<number, IStaff>,
}

export default class Company implements ICompany {
  public companyName: string;

  private root: ITeam;

  constructor(name: string) {
    this.companyName = name;
    this.root = new Team(name);
  }

  public getRoot(): ITeam {
    // Returns iterator for the values of the base node. 
    // It only ever has one node, which is the top level of the company
    return this.root;

  }

  public addTeam(team: ITeam, ancestors: string[]): Map<string, ITeam> {
    // Searches through list of ancestor teams back to root to make sure team is being added to connected tree
    // If is the first child then appends directly to root company node
    if(!this.root.getChildren().size) {
      return this.root.addChild(team);
    }

    const parent = this.search(ancestors);
    team.setParent(parent);
    return parent.addChild(team);
  }

  public search(pathToTeam: string[]): ITeam {
    // Searches for nodes by checking each nodes Map() of its children for the next value in the list
    // If found it returns the object and then checks that object children for the next node in the path
    let currentAncestor: ITeam = this.root;
    let nextAncestor = pathToTeam.shift();

    while(nextAncestor && currentAncestor.hasChild(nextAncestor)) {
      currentAncestor = currentAncestor.getChild(nextAncestor);
      if (!pathToTeam.length) {
        return currentAncestor;
      }
      nextAncestor = pathToTeam.shift();
    }
    throw new Error('Not found');
  }

  public addMemberToTeam(pathToTeam: string[], staff: IStaff): Map<number, IStaff> {
    // Rehashed search function adds staff member to each node enroute to finding the node representing the team the staff member belongs to
    // Can help avoid expense of having to visit each nested node later
    let currentAncestor: ITeam = this.root;
    let nextAncestor = pathToTeam.shift();

    while(nextAncestor && currentAncestor.hasChild(nextAncestor)) {
      currentAncestor = currentAncestor.getChild(nextAncestor);
      // Had trouble merging maps so adding immediate and nested staff to nested staff Map() until time for a fix
      currentAncestor.addNestedStaff(staff);

      if (!pathToTeam.length) {
        currentAncestor.addStaff(staff);
        return currentAncestor.getStaff();
      }

      nextAncestor = pathToTeam.shift();
    }
    throw new Error('Not found');
  }

  public getStaff(pathToTeam: string[]): Map<number, IStaff> {
    return this.search(pathToTeam).getNestedStaff();
    
  }
}
