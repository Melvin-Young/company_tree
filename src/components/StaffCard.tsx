/* tslint:disable no-unused-expression jsx-no-lambda */

import React from "react";
import { IStaff } from "../utils/Team";

interface IProps {
  staff: IStaff;
}

const StaffCard = (props: IProps) => (
  <div className="card">
    <img
      src={require("./../images/profile.png")}
      alt="Avatar"
      className="staff-picture"
    />
    <h2> {props.staff.name}</h2>
  </div>
);

export default StaffCard;