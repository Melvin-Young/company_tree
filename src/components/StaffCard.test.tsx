/* tslint:disable no-unused-expression jsx-no-lambda */

import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';

import { IStaff } from '../utils/Team';
import StaffCard from './StaffCard';

Enzyme.configure({ adapter: new Adapter()});

describe('StaffCard', () => {
    it('should render', () => {
        const component = shallow(<StaffCard staff={{} as IStaff}/>)
        expect(component).toMatchSnapshot();
    });
});