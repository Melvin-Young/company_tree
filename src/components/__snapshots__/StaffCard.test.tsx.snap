// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaffCard should render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StaffCard
    staff={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <img
          alt="Avatar"
          className="staff-picture"
          src="profile.png"
/>,
        <h2>
           
</h2>,
      ],
      "className": "card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "alt": "Avatar",
          "className": "staff-picture",
          "src": "profile.png",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            " ",
            undefined,
          ],
        },
        "ref": null,
        "rendered": Array [
          " ",
          undefined,
        ],
        "type": "h2",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <img
            alt="Avatar"
            className="staff-picture"
            src="profile.png"
/>,
          <h2>
             
</h2>,
        ],
        "className": "card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "alt": "Avatar",
            "className": "staff-picture",
            "src": "profile.png",
          },
          "ref": null,
          "rendered": null,
          "type": "img",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              " ",
              undefined,
            ],
          },
          "ref": null,
          "rendered": Array [
            " ",
            undefined,
          ],
          "type": "h2",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
